OpenRA.Mods.Common/Traits/Player/ParallelProductionQueue.cs

```diff
@@ -0,0 +1,70 @@
+﻿using System.Collections.Generic;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class ParallelProductionQueueInfo : ProductionQueueInfo
+	{
+		public override object Create(ActorInitializer init) { return new ParallelProductionQueue(init, this); }
+	}
+
+	public class ParallelProductionQueue : ProductionQueue
+	{
+		[Sync] public new bool CurrentDone { get { return Queue.Find(i => i.Done) != null; } }
+
+		public ParallelProductionQueue(ActorInitializer init, ProductionQueueInfo info) : base(init, init.Self.Owner.PlayerActor, info)
+		{
+		}
+
+		protected override void ProgressQueue()
+		{
+			for (var i = 0; i < Queue.Count; i++)
+			{
+				var first = Queue[0];
+
+				var before = first.RemainingTime;
+				first.Tick(playerResources);
+
+				foreach (var item in Queue.FindAll(a => a.Item == first.Item))
+				{
+					Queue.Remove(item);
+					Queue.Add(item);
```

Comment:
Are you sure this is working as intended because you are reordering the queue as you are iterating over it.
I also wanted to ask about why you don't use a foreach at the top but that would probably throw an exception with the add and remove.

Reply:
reordering is important, otherwise if you have low money, the first producting item will consume all the money, and the others will not progress at all. it just makes sure on the next tick, the next item will try to progress.

Reply:
Reordering while looping through the collection is still bad, you could be ticking the same item twice for example.